### ISSUE TITLE
Question: Decoding base64 encrypted strings.

### ISSUE THREAD

**Issue by VictorUvarov** [leocavalcante/encrypt#69]
I have an encrypted string and key, for example: 

cipherText =`Gxbx6Q74Yho6dcGH+5EgT0cyV1M0dUZjck9pMVJRNk9EeStUa2c9PQ==`
Secret Key = `'88aeb0df37b2b028779df80a92815df5'`

The cipherText is encrypted with `AES-128-CBC`, has an IV of length 16,  and a secret key that is hashed with md5 and prepended to the cipherText before being encoded in base64.

Psuedo Code: `ciphertext = base64_encode( iv + ciphertext_raw )`

The code I have is complaining that:
`Unhandled Exception: Invalid argument(s): Initialization vector must be the same length as block size`.

```dart
String aesDecrypt(String encryptedString, String sharedSecretHash) {
  // Decode the cipherText and convert to string
  Uint8List decodedL = base64.decode(encryptedString);
  String decoded = String.fromCharCodes(decodedL);

  // Extract the iv and cipherText from the
  String ivDecoded = decoded.substring(0, 16);
  String cipherTextDecoded = decoded.substring(16);

  // Setup key and encrypter to use AES-CBC
  final key = Key.fromUtf8(sharedSecretHash);
  final encrypter = Encrypter(AES(key, mode: AESMode.cbc));

  // re encode the iv so that is can be parsed by IV
  final iv = IV.fromUtf8(ivDecoded);

  // Decrypt cipherText
  final encrypted = Encrypted.fromUtf8(cipherTextDecoded);
  final originalPlainText = encrypter.decrypt(encrypted, iv: iv);

  return originalPlainText;
}
```

Not sure what I'm supposed to do

**Comment by VictorUvarov** [comment]
For some reason `final iv = IV.fromUtif8(ivDecoded);` turns the length of iv into 23 instead of keeping it at 16

**Comment by leocavalcante** [comment]
I didn't get how was the IV and the encrypted text concatenated `iv + ciphertext_raw`. There is a token between? Was just the bytes appended right after the other? How they was represented before the concatenation? UTF8, base 64, hex, raw bytes?

Also, isn't your key hexdecimal represented instead of an UTF-8 string?

**Comment by VictorUvarov** [comment]
I was concatenatng the iv and the cipher text before encoding to base64. The inputs were just strings. Here is the example code in `php`

openssl_encrypt docs: https://www.php.net/manual/en/function.openssl-encrypt.php
openssl_random_pseudo_bytes docs: https://www.php.net/manual/en/function.openssl-random-pseudo-bytes.php
```php
$cipher = "AES-128-CBC";
$hashAlgo = 'md5'
$ivLength = 16
$sharedSecret = 'SHAREDSECRET';
$plaintext = 'HELLO WORLD';
$sharedSecretHash = hash($hashAlgo, $sharedSecret);
$encrypted = aes_encrypt($plaintext, $sharedSecretHash, $cipher);
$decrypted = aes_decrypt($encrypted, $sharedSecretHash, $cipher);

function aes_encrypt($plaintext, $sharedSecretHash, $cipher)
{
   $iv             = openssl_random_pseudo_bytes($ivLength);
   $ciphertext_raw = openssl_encrypt($plaintext, $cipher, $sharedSecretHash, 0, $iv);
   return $ciphertext = base64_encode( $iv.$ciphertext_raw );
}

function aes_decrypt($ciphertext, $sharedSecretHash, $cipher)
{
   $c     = base64_decode($ciphertext);
   $iv    = substr($c, 0, $ivLength);
   $ciphertext_raw = substr($c, $ivLength);
   $original_plaintext = openssl_decrypt($ciphertext_raw, $cipher, $sharedSecretHash, 0, $iv);
   return $original_plaintext;

}
```

**Comment by leocavalcante** [comment]
# tl;dr:
Got this working with the following snippets:
###### encrypt.php
```php
<?php declare(strict_types=1);

$data = 'Hello World';
$method = 'AES-256-CBC';
$key = base64_decode('ABYYFR0gGBoHGwcDCB0NGxgEDgkbGhQGBQMdCyAfARw=');
$iv = openssl_random_pseudo_bytes(16);
$encrypted = openssl_encrypt($data, $method, $key, OPENSSL_RAW_DATA, $iv);

echo base64_encode($iv.$encrypted);
```

###### decrypt.dart
```dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:crypto/crypto.dart';
import 'package:encrypt/encrypt.dart';

void main(List<String> args) {
  final token = base64.decode(args[0]);
  final key = Key.fromBase64('ABYYFR0gGBoHGwcDCB0NGxgEDgkbGhQGBQMdCyAfARw=');
  final iv = IV(Uint8List.fromList(token.sublist(0, 16)));
  final encrypted = Encrypted(Uint8List.fromList(token.sublist(16)));
  final encrypter = Encrypter(AES(key, mode: AESMode.cbc));

  print(encrypter.decrypt(encrypted, iv: iv));
}

String hexString(String input) => md5.convert(utf8.encode(input)).toString();
```

###### Running
```pwsh
issue_69 [master]~ dart .\decrypt.dart (php .\encrypt.php)
Hello World
```

# What is happening
Actually, a lot is going on
(and a just learn more about it, thank you for bringing this up).

The first thing I noticed is that using the value `0` on the `$options` parameter of the `openssl_encrypt()` function, makes your output be already a base 64 encoded string (that is why I use `OPENSSL_RAW_DATA` on the snippet).

Another thing is that your key being a MD5 hash of the `SHAREDSECRET` string. This isn't secure, you should rely on [cryptographically secure pseudorandom number generators ](https://en.wikipedia.org/wiki/Cryptographically_secure_pseudorandom_number_generator). Its implemented on the library, you can [create one on the CLI](https://github.com/leocavalcante/encrypt#secure-random). Also, the default value of the third param of the `hash()` function is `FALSE` which means that no raw output will be returned, you get a hexdecimal string representation instead, so you are actually using the hexdecimal representation of the bytes as an UTF-8/ASCII string, not the actual bytes of the hashing result.

The last thing was what got me for a few hours 😌.
**PHP crops the key to 16 bytes length!** Using AES-**128**-CBC makes PHP to crop and use only the first 16 bytes (because 128 / 8) of the key, so even if you are passing a 32 bytes length string (the result of MD5) PHP is actually using only the first half of it. So o probably noticed that I used `AES-256-CBC` on the snippets to use a 32 bytes length string...

## So, to make your decryption work using what you are doing in PHP:
```php
<?php

$cipher = "AES-128-CBC";
$hashAlgo = 'md5';
$ivLength = 16;
$sharedSecret = 'SHAREDSECRET';
$plaintext = 'HELLO WORLD';
$sharedSecretHash = hash($hashAlgo, $sharedSecret);
$encrypted = aes_encrypt($plaintext, $sharedSecretHash, $cipher);
// $decrypted = aes_decrypt($encrypted, $sharedSecretHash, $cipher);

echo $encrypted;

function aes_encrypt($plaintext, $sharedSecretHash, $cipher)
{
  global $ivLength;

   $iv             = openssl_random_pseudo_bytes($ivLength);
   $ciphertext_raw = openssl_encrypt($plaintext, $cipher, $sharedSecretHash, 0, $iv);
   return $ciphertext = base64_encode( $iv.$ciphertext_raw );
}

function aes_decrypt($ciphertext, $sharedSecretHash, $cipher)
{
   $c     = base64_decode($ciphertext);
   $iv    = substr($c, 0, $ivLength);
   $ciphertext_raw = substr($c, $ivLength);
   $original_plaintext = openssl_decrypt($ciphertext_raw, $cipher, $sharedSecretHash, 0, $iv);
   return $original_plaintext;
}
```
You have to (code comments):
```dart
import 'dart:convert';

import 'dart:typed_data';

import 'package:crypto/crypto.dart';
import 'package:encrypt/encrypt.dart';

void main(List<String> args) {
  // Decode the input
  final token = base64.decode(args[0]);

  // Get the IV from the first 16 bytes
  final iv = IV(Uint8List.fromList(token.sublist(0, 16)));

  // Get the encrypted string from a base64 string of the second part, because $options is 0 on PHP
  final encrypted = Encrypted.fromBase64(utf8.decode(token.sublist(16)));

  // Use substring(0, 16) on the hexdecimal representation of the string and using as UTF-8 instead of actually using the bytes because the third param on hash() of PHP is FALSE
  final key = Key.fromUtf8(hexString('SHAREDSECRET').substring(0, 16));

  // Then it should work
  final encrypter = Encrypter(AES(key, mode: AESMode.cbc));
  final decrypted = encrypter.decrypt(encrypted, iv: iv);
  print(decrypted);
}

String hexString(String input) => md5.convert(utf8.encode(input)).toString();
```
```pwsh
issue_69 [master]~ dart .\issue.dart (php .\issue.php)
HELLO WORLD
```

**Comment by VictorUvarov** [comment]
wow thanks a lot for the detailed explanation. I'll look into those secure PHP changes you suggested.
